src/Compilers/CSharp/Portable/Binder/Binder_TupleOperators.cs

```diff
@@ -0,0 +1,369 @@
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal partial class Binder
+    {
+        /// <summary>
+        /// If the left and right are tuples of matching cardinality, we'll try to bind the operator element-wise.
+        /// When that succeeds, the element-wise conversions are collected and their result type is applied as tuple conversions.
+        /// The element-wise binary operators are collected and stored as a tree for lowering.
+        /// </summary>
+        private BoundTupleBinaryOperator BindTupleBinaryOperator(BinaryExpressionSyntax node, BinaryOperatorKind kind,
+            BoundExpression left, BoundExpression right, DiagnosticBag diagnostics)
+        {
+            TupleBinaryOperatorInfo operators = BindTupleBinaryOperatorInfo(node, kind, left, right, diagnostics);
+
+            BoundExpression convertedLeft = GenerateConversionForAssignment(operators.LeftConvertedType, left, diagnostics);
+            BoundExpression convertedRight = GenerateConversionForAssignment(operators.RightConvertedType, right, diagnostics);
+
+            TypeSymbol resultType = GetSpecialType(SpecialType.System_Boolean, diagnostics, node);
+            return new BoundTupleBinaryOperator(node, convertedLeft, convertedRight, kind, operators, resultType);
+        }
+
+        private TupleBinaryOperatorInfo BindTupleBinaryOperatorInfo(BinaryExpressionSyntax node, BinaryOperatorKind kind,
+            BoundExpression left, BoundExpression right, DiagnosticBag diagnostics)
+        {
+            TypeSymbol leftType = left.Type;
+            TypeSymbol rightType = right.Type;
+            int? leftCardinality = GetTupleCardinality(left);
+            int? rightCardinality = GetTupleCardinality(right);
+
+            if (leftCardinality.HasValue && rightCardinality.HasValue)
+            {
+                if (leftCardinality.Value != rightCardinality.Value)
+                {
+                    Error(diagnostics, ErrorCode.ERR_TupleSizesMismatchForBinOps, node, leftCardinality, rightCardinality);
+
+                    return new TupleBinaryOperatorInfo.Single(leftType ?? CreateErrorType(), rightType ?? CreateErrorType(),
+                        BinaryOperatorKind.Error, methodSymbolOpt: null, boolConversion: Conversion.NoConversion, boolOperator: default);
+                }
+
+                // typeless tuple literals are not nullable
+                bool leftNullable = leftType?.IsNullableType() == true;
+                bool rightNullable = rightType?.IsNullableType() == true;
+
+                return BindTupleBinaryOperatorNestedInfo(node, kind,
+                    GetTupleArgumentsOrPlaceholders(left), GetTupleArgumentsOrPlaceholders(right),
+                    leftNullable || rightNullable, diagnostics);
+            }
+
+            return BindTupleBinaryOperatorSingleInfo(node, kind, left, right, diagnostics);
```

Comment:
>return [](start = 12, length = 6)

I think this is unreachable.

Reply:
There are two callers of `BindTupleBinaryOperatorInfo`. One is `BindTupleBinaryOperator` (entry point, line 20) and the other is recursive case (`BindTupleBinaryOperatorNestedInfo` line 207).

---
In reply to: [168054713](https://github.com/dotnet/roslyn/pull/23356#discussion_r168054713) [](ancestors = 168054713)